### ISSUE TITLE
subprojects regression / proper subproject reusage?

### ISSUE THREAD

**Issue by m1027** [mesonbuild/meson#3155]
I have a directory layout to *reuse subprojects from multiple projects*:

projects/proj1 (may use foo/v2)
projects/proj2 (may use foo/v2 + baz/v1)
projects/proj3 (may use foo/v1, ...)

utilities/foo/v1
utilities/foo/v2
utilities/bar/
utilities/baz/v1
utilities/baz/v2

But after an upgrade from 0.41 to 0.44, and especially subprojects_dir being
deprecated now, the layout for subprojects seems to be:

* proj1/subprojects/...
* proj2/subprojects/...
* proj3/subprojects/...

However, this requires to have each project have its own set of subprojects,
without sharing.

My current workaround is using ugly symlinks:

* proj1/subprojects/foo (=../../utilities/foo/v1)
* proj1/subprojects/baz (=../../utilities/baz/v2)
* ...

What is the recommended approach to this?

Notes on my requirements:

* Each project shall use/call any of the utility's (subproject's) meson.build
  files.

* Build details of each utility shall be kept in it's own meson.build file
  rather than copying details into proj1, proj2, proj3's meson.build file.

* Optionally compile each utility stand alone with non-subproject options
  in its meson.build file (by checking 'is_subproject').

Thanks

**Comment by jpakkane** [comment]
If you do this:

    root
      subprojects
        foo-v1
        foo-v2
        bar-v1
        bar-v2

Then you can do `subproject('foo-v1')` in any other subproject and it works. Subprojects are always searched for in the top level subproject directory.

**Comment by m1027** [comment]
> If you do this:
> 
> root
>   subprojects
>     foo-v1
>     foo-v2
>     bar-v1
>     bar-v2
> 
> Then you can do subproject('foo-v1') in any other subproject and it works.
> Subprojects are always searched for in the top level subproject directory.

Thanks but my setup is different:

```
root01
  subprojects
    foo-v1
    foo-v2
    bar-v1
    bar-v2
```

and

```
root02
  subprojects
    foo-v1
    foo-v2
```

As you can see root02 wants to reuse some of the subprojects from root01.

I even prefer to have subprojects like a pool of utilites outside of the 
root projects.

In meson 0.41 I achieved that by something like

* subprojects_dir = ../../utilities/
* and using subprojects('foo/v1') etc.

But now, subprojects_dir and subprojects() are not allowed to be neither
absolute paths nor relative ones...

So to sum up:

I just ask for the recommended meson way to re-use subprojects (helper code,
libraries etc) across multiple root projects.

**Comment by nirbheek** [comment]
fwiw, I use symlinks to share libraries between projects. I don't think a dedicated mechanism to do this is necessarily needed in meson, it would add complexity and more failure conditions for little benefit.

You can also use symlinks on Windows (`mklink`), you just need administrator privileges, or you can edit a regedit key to enable it for everyone.

**Comment by nirbheek** [comment]
> I seem to miss the conceptual reason and benefit why to force subprojects into a subdirectory of the root project

The point is that source trees should be self-contained. Everything you need is inside it, that way if you change something in a subproject, you know it won't inadvertantly break some other project.

Finally, the only thing you will share in the end is the sources, and meson does not really do source management. It will download a subproject for you, but it won't, for instance, update it for you.

**Comment by m1027** [comment]
> The point is that source trees should be self-contained. Everything you need
> is inside it, that way if you change something in a subproject, you know it
> won't inadvertantly break some other project.

Self-containment is a weak approach of not breaking things. It may help in
some areas but in others it doesn't: When you start copying and pasting
identical subprojects (source code) into multiple root projects then bug
fixing becomes a mess.

One might argue to create libraries or even packages for subprojects in such
cases. But think of, say, a widget library wrapper you wrote and want to use
in every root project: It may still not be worth being a separate library nor
worth bothering packet managers and maintainers with it.

So, I definitely prefer keeping my helper code in a single pool, accessed and
reused by all root projects!


> Finally, the only thing you will share in the end is the sources, and meson
> does not really do source management. It will download a subproject for you,
> but it won't, for instance, update it for you.

Hm. If meson were not about source management it should not force to keep the
sources managed in a self-contained directory.

Hey, as long as there are no other reasons, I'd really appreciate if meson
would simply reenable the option to call subprojects even if located somewhere
outside the root project.
